src/lib.rs

```diff
@@ -1015,6 +1015,100 @@ impl Rng for ThreadRng {
     }
 }
 
+/// An RNG provided specifically for seeding PRNG's. It uses the interface for
+/// random numbers provided by the operating system (`OsRng`). If that returns
+/// an error, it will fall back to the `JitterRng` entropy collector.
+/// Occasionally it will then check if `OsRng` is still not available.
+#[derive(Debug)]
+pub struct EntropySource {
+    rng: EntropySourceInner,
+    counter: u32,
+}
+
+#[derive(Debug)]
+enum EntropySourceInner {
+    Os(OsRng),
+    Jitter(JitterRng),
+}
+
+impl EntropySource {
+    pub fn new() -> Result<Self, Error> {
+        match OsRng::new() {
+            Ok(r) =>
+                Ok(EntropySource { rng: EntropySourceInner::Os(r),
+                                   counter: 0u32 }),
+            Err(e1) => {
+                match JitterRng::new() {
+                    Ok(r) =>
+                        Ok(EntropySource { rng: EntropySourceInner::Jitter(r),
+                                           counter: 0 }),
+                    Err(_) =>
+                        Err(Error::with_cause(
+                            ErrorKind::Unavailable,
+                            "Both OS and Jitter entropy sources are unavailable",
+                            e1))
+                }
+            }
+        }
+    }
+}
+
+impl Rng for EntropySource {
+    fn next_u32(&mut self) -> u32 {
+        impls::next_u32_via_fill(self)
+    }
+
+    fn next_u64(&mut self) -> u64 {
+        impls::next_u64_via_fill(self)
+    }
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        self.try_fill_bytes(dest).unwrap();
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+        let mut switch_rng = None;
+        let mut result;
+        match self.rng {
+            EntropySourceInner::Os(ref mut rng) => {
+                result = rng.try_fill_bytes(dest);
+                if result.is_err() {
+                    // Fallback to JitterRng.
+                    let mut rng2_result = JitterRng::new();
+                    if let Ok(mut rng2) = rng2_result {
+                        result = rng2.try_fill_bytes(dest); // Can't fail
+                        switch_rng = Some(EntropySourceInner::Jitter(rng2));
```

Comment:
Probably simpler to recurse here (i.e. call `new_self.try_fill_bytes()`).

Reply:
It took some thinking to get the logic right and the borrow checker happy. But I'll think about about it!